Don't initialize signal handing machinery if we're not going to use it.

MenloDorian opened the pull request:
#### 0051847d2a3954b88fc4669e9238f730df68cfd4
<pre>
Don&apos;t initialize signal handing machinery if we&apos;re not going to use it.
<a href="https://bugs.webkit.org/show_bug.cgi?id=252651">https://bugs.webkit.org/show_bug.cgi?id=252651</a>
rdar://105712893

Reviewed by Yusuke Suzuki.

This saves some memory when the signal handling machinery is not needed.

1. Move the handling of JS_SHELL_WAIT_FOR_SIGUSR2_TO_EXIT in jsc.cpp into jscmain.
   This is because addSignalHandler() depends on JSC Options, and the Options have
   not be initialized yet at its previous location.

2. Make signal handling stages explicit with an InitState in WTF::Config.  The
   InitState can go from Uninitialized to InitializedHandlerThread, and then to
   AddedHandlers.  The InitState cannot go in reverse.

   The client must call initializeSignalHandling() to advance the InitState from
   Uninitialized to InitializedHandlerThread.

   The client must have called addSignalHandler() in order to advance the InitState
   from InitializedHandlerThread to AddedHandlers.  If addSignalHandler() is called
   without first calling initializeSignalHandling(), a RELEASE_ASSERT will fail.

   The client can only call registerThreadForMachExceptionHandling() after one or
   more addSignalHandler() has been called.  registerThreadForMachExceptionHandling()
   will not RELEASE_ASSERT that the InitState is AddedHandlers.

3. Added stub functions for Wasm::isSupported(), Wasm::activateSignalingMemory(),
   Wasm::prepareSignalingMemory(), Wasm::startTrackingCurrentThread(), and
   Wasm::resetInstructionCacheOnAllThreads() so that they can be called without first
   checking #if on some WASM #define variable.

* Source/JavaScriptCore/jsc.cpp:
(main):
(CommandLine::parseArguments):
(jscmain):
* Source/JavaScriptCore/runtime/InitializeThreading.cpp:
(JSC::initialize):
* Source/JavaScriptCore/runtime/VMEntryScope.cpp:
(JSC::VMEntryScope::VMEntryScope):
* Source/JavaScriptCore/tools/JSDollarVM.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/wasm/WasmCapabilities.h:
(JSC::Wasm::isSupported):
* Source/JavaScriptCore/wasm/WasmFaultSignalHandler.h:
(JSC::Wasm::activateSignalingMemory):
(JSC::Wasm::prepareSignalingMemory):
* Source/JavaScriptCore/wasm/WasmMachineThreads.h:
(JSC::Wasm::startTrackingCurrentThread):
(JSC::Wasm::resetInstructionCacheOnAllThreads):
* Source/WTF/wtf/threads/Signals.cpp:
(WTF::SignalHandlers::add):
(WTF::initMachExceptionHandlerThread):
(WTF::registerThreadForMachExceptionHandling):
(WTF::addSignalHandler):
(WTF::activateSignalHandlersFor):
(WTF::startMachExceptionHandlerThread): Deleted.
* Source/WTF/wtf/threads/Signals.h:
(WTF::initializeSignalHandling):
(WTF::disableSignalHandling):
(WTF::registerThreadForMachExceptionHandling):
* Tools/TestWebKitAPI/Tests/WTF/Signals.cpp:
(TEST):

Canonical link: <a href="https://commits.webkit.org/260623@main">https://commits.webkit.org/260623@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d1cae4adc610dadf1d53c59c6f5ce70269dad40a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108837 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17939 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41672 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/372 "Built successfully") | [❌ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118124 "Hash d1cae4ad for PR 10431 does not build (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112718 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19395 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9218 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101071 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114605 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/14560 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97759 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42531 "Failed to checkout and rebase branch from PR 10431") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96529 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29399 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/84368 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/97931 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10715 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30748 "Passed tests") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/38/builds/98739 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/8842 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11470 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7682 "Passed tests") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/38/builds/98739 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16859 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50347 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/106363 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7340 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13057 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26367 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/jsc.cpp:
```diff
@@ -3072,6 +3072,8 @@ JSC_DEFINE_HOST_FUNCTION(functionDropAllLocks, (JSGlobalObject* globalObject, Ca
 #define EXCEPT(x)
 #endif
 
+static LazyNeverDestroyed<BinarySemaphore> waitToExit;
```
You can put `static BinarySemaphore waitToExit` (`LazyNeverDestroyed` is not necessary).

Constellation commented on Source/JavaScriptCore/runtime/InitializeThreading.cpp:
```diff
@@ -119,14 +119,16 @@ void initialize()
-#if ENABLE(WEBASSEMBLY)
-        Wasm::prepareSignalingMemory();
-#endif
+        if (Wasm::isSupported() || !Options::usePollingTraps()) {
```
Should we make `Options::useWebAssembly()` work instead and remove `Wasm::isSupported`?
(putting `false` to `Options::useWebAssembly()` in Options.cpp when it is not supported etc., like `Options::useJIT()`).

MenloDorian replied:
We already disable `Options::useWebAssembly()` in `disableAllJITOptions()`.  I think it's good to replace `Wasm::isSupported()` with `Options::useWebAssembly()`, but I'll do that in a follow up patch.

Constellation commented on Source/JavaScriptCore/runtime/InitializeThreading.cpp:
```diff
@@ -119,14 +119,16 @@ void initialize()
+
+            if (!Options::usePollingTraps())
+                VMTraps::initializeSignals();
+            if (Wasm::isSupported())
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/VMEntryScope.cpp:
```diff
@@ -47,12 +48,10 @@ VMEntryScope::VMEntryScope(VM& vm, JSGlobalObject* globalObject)
-                    Wasm::startTrackingCurrentThread();
-#endif
-
+            if (Wasm::isSupported())
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmCapabilities.h:
```diff
@@ -31,9 +31,15 @@
 namespace JSC {
 namespace Wasm {
 
+#if ENABLE(WEBASSEMBLY)
+
 inline bool isSupported()
 {
     return Options::useWebAssembly();
 }
 
+#else
+inline bool isSupported() { return false; }
+#endif
```
Ditto.

Constellation reviewed (approved):
r=me with comment.

Commit:
Don't initialize signal handing machinery if we're not going to use it.
https://bugs.webkit.org/show_bug.cgi?id=252651
rdar://105712893

Reviewed by Yusuke Suzuki.

This saves some memory when the signal handling machinery is not needed.

1. Move the handling of JS_SHELL_WAIT_FOR_SIGUSR2_TO_EXIT in jsc.cpp into jscmain.
   This is because addSignalHandler() depends on JSC Options, and the Options have
   not be initialized yet at its previous location.

2. Make signal handling stages explicit with an InitState in WTF::Config.  The
   InitState can go from Uninitialized to InitializedHandlerThread, and then to
   AddedHandlers.  The InitState cannot go in reverse.

   The client must call initializeSignalHandling() to advance the InitState from
   Uninitialized to InitializedHandlerThread.

   The client must have called addSignalHandler() in order to advance the InitState
   from InitializedHandlerThread to AddedHandlers.  If addSignalHandler() is called
   without first calling initializeSignalHandling(), a RELEASE_ASSERT will fail.

   The client can only call registerThreadForMachExceptionHandling() after one or
   more addSignalHandler() has been called.  registerThreadForMachExceptionHandling()
   will not RELEASE_ASSERT that the InitState is AddedHandlers.

3. Added stub functions for Wasm::isSupported(), Wasm::activateSignalingMemory(),
   Wasm::prepareSignalingMemory(), Wasm::startTrackingCurrentThread(), and
   Wasm::resetInstructionCacheOnAllThreads() so that they can be called without first
   checking #if on some WASM #define variable.

* Source/JavaScriptCore/jsc.cpp:
(main):
(CommandLine::parseArguments):
(jscmain):
* Source/JavaScriptCore/runtime/InitializeThreading.cpp:
(JSC::initialize):
* Source/JavaScriptCore/runtime/VMEntryScope.cpp:
(JSC::VMEntryScope::VMEntryScope):
* Source/JavaScriptCore/tools/JSDollarVM.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/wasm/WasmCapabilities.h:
(JSC::Wasm::isSupported):
* Source/JavaScriptCore/wasm/WasmFaultSignalHandler.h:
(JSC::Wasm::activateSignalingMemory):
(JSC::Wasm::prepareSignalingMemory):
* Source/JavaScriptCore/wasm/WasmMachineThreads.h:
(JSC::Wasm::startTrackingCurrentThread):
(JSC::Wasm::resetInstructionCacheOnAllThreads):
* Source/WTF/wtf/threads/Signals.cpp:
(WTF::SignalHandlers::add):
(WTF::initMachExceptionHandlerThread):
(WTF::registerThreadForMachExceptionHandling):
(WTF::addSignalHandler):
(WTF::activateSignalHandlersFor):
(WTF::startMachExceptionHandlerThread): Deleted.
* Source/WTF/wtf/threads/Signals.h:
(WTF::initializeSignalHandling):
(WTF::disableSignalHandling):
(WTF::registerThreadForMachExceptionHandling):
* Tools/TestWebKitAPI/Tests/WTF/Signals.cpp:
(TEST):

Canonical link: https://commits.webkit.org/260623@main
